x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -232,19 +235,136 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
         }
     }
 
-    private XContentBuilder resolveEnrichMapping(final EnrichPolicy policy) {
-        // Currently the only supported policy type is EnrichPolicy.MATCH_TYPE, which is a keyword type
-        final String keyType;
-        final CheckedFunction<XContentBuilder, XContentBuilder, IOException> matchFieldMapping;
+    private void resolveEnrichMapping(
+        final EnrichPolicy policy,
+        GetIndexResponse getIndexResponse,
+        ActionListener<XContentBuilder> resultListener
+    ) {
         if (EnrichPolicy.MATCH_TYPE.equals(policy.getType())) {
-            matchFieldMapping = (builder) -> builder.field("type", "keyword").field("doc_values", false);
-            // No need to also configure index_options, because keyword type defaults to 'docs'.
+            createEnrichMappingBuilder((builder) -> builder.field("type", "keyword").field("doc_values", false), resultListener);
+        } else if (EnrichPolicy.RANGE_TYPE.equals(policy.getType())) {
+            resolveRangeTypeEnrichMapping(policy, getIndexResponse, resultListener);
         } else if (EnrichPolicy.GEO_MATCH_TYPE.equals(policy.getType())) {
-            matchFieldMapping = (builder) -> builder.field("type", "geo_shape");
+            createEnrichMappingBuilder((builder) -> builder.field("type", "geo_shape"), resultListener);
         } else {
-            throw new ElasticsearchException("Unrecognized enrich policy type [{}]", policy.getType());
+            resultListener.onFailure(new ElasticsearchException("Unrecognized enrich policy type [{}]", policy.getType()));
         }
+    }
 
+    private void resolveRangeTypeEnrichMapping(
+        EnrichPolicy policy,
+        GetIndexResponse response,
+        ActionListener<XContentBuilder> resultListener
+    ) {
+        String matchField = policy.getMatchField();
+        List<String> types = collapseFieldMappingsValue(response, toFieldMappingProperty(matchField, "type"));
+        if (types.isEmpty()) {
```

Comment:
I think with the above simplification (using `ObjectPath#eval(...)`), this can just be a null check? (instead of checking for empty and more than size 1 (which shouldn't be possible?))

Reply:
Similar to below, a user is currently allowed to specify multiple source indicies which might have conflicting mappings.